#137005167 paystack payment and wallet

vonvick opened the pull request:
#### What does this PR do?
This PR integrates Paystack payment processor as one of the payment methods for the application.
It also integrates wallet functionality that can be funded by paystack and funds can be transferred from one consumer to the other.
#### Description of Task to be completed?
Integrate paystack as one of the payment options.
Wallets.
Consumer to consumer funds transfer.
#### How should this be manually tested?
this can be tested by buying a product and proceeding to paystack to make payments where the payment methods can be used to make payments
#### Any background context you want to provide?
Given that a user has added a product to cart and wants to check out, they can see the option to pay with paystack or with their wallet.
The user should also be able to transfer funds to other users of the application.
#### What are the relevant pivotal tracker stories?
#137005167
#### Screenshots (if appropriate)
![paystack-payment](https://cloud.githubusercontent.com/assets/21035558/22514152/afced35c-e89e-11e6-8986-123d99966bf4.png)
![wallet-dashboard1](https://cloud.githubusercontent.com/assets/21035558/22514154/afd212ec-e89e-11e6-99e9-f229fb896dee.png)
![wallet-dashboard2](https://cloud.githubusercontent.com/assets/21035558/22514153/afd04002-e89e-11e6-90a5-5f507e6de354.png)
![wallet-dropdown](https://cloud.githubusercontent.com/assets/21035558/22514151/afcecbd2-e89e-11e6-8b71-fcf334e1f93c.png)

#### Questions:
Not Applicable

Commit:
feature(currency): add naira currency to shop.json

Commit:
feature(paystack): add paystack javascript library to browser policy.

Commit:
feature(paystack): add paystack library to index.html

Commit:
feature(paystack): create register for paystack payment method

Commit:
feature(paystack): create template for paystack

create template and helper files for paystack checkout and settings(dashboard) layout.

[Start #137005167]

Commit:
feature(paystack): create schema for paystack

create schema to store paystack package keys and information

[WIP #137005167]

Commit:
feature(paystack): fix issue with  paystack public key

Commit:
feature(wallet): add wallet option to user's account page

Commit:
feature(wallet): add route to wallet on shop object

Commit:
feature(wallet): add wallet option to checkout page

Commit:
feature(wallet): create wallet transaction form on user's account page

Commit:
feature(wallet): create wallet collections and link up with user's account

Commit:
feature(wallet): add wallet to default roles for registered users

Commit:
feature(wallet): create methods for wallet

create Meteor methods for wallet collection to credit, debit and refund the user

[Finishes #137005167]

Commit:
style(eslint): fix issue with consistent return on files

andela-aolaniran commented on imports/plugins/core/layout/client/templates/layout/wallet/wallet.js:
```diff
@@ -0,0 +1,144 @@
+Template.wallet.onCreated(function bodyOnCreated() {
+  this.state = new ReactiveDict();
+  this.autorun(() => {
+    this.state.setDefault({
```
Setting your default state within the autorun() would lead to an unnecessary continuous resetting of the default state when a reactive state changes.

vonvick replied:
I agree with you and it has been fixed

Commit:
style(eslint): fix issues with houndci while raising a PR

andela-aolaniran commented on imports/plugins/included/paystack/client/settings/paystack.js:
```diff
@@ -0,0 +1,31 @@
+    },
+    onError: function (operation, error) {
+      Alerts.removeSeen();
+      return Alerts.add("Paystack settings update failed. " + error, "danger");
```
Can we use a template literal here.

olawalejarvis replied:
I agree with @andela-aolaniran  since that's the style they adopted in the code base

vonvick replied:
it has been fixed

Commit:
bug(wallet): fix issue with wallet not updating

Commit:
style(eslint): edit eslint file to fix hound issues

Commit:
style(eslint): fix more issues with houndci

Commit:
bug(wallet): fix error with shops not defined.

Commit:
style(eslint): fix more houndci issues

olawalejarvis commented on server/methods/core/wallet.js:
```diff
@@ -0,0 +1,75 @@
+Meteor.publish("transactionDetails", (userId) => {
+  check(userId, String);
+  return Wallets.find({userId});
+});
```
I think it is better to put publication code inside the publication folder

vonvick replied:
fixed

Commit:
refactor(payment): refactor some code

refactor and make corrections to code based off of review received for the PR

[Fixes  #137005167]

olawalejarvis reviewed (approved):
Approved

Commit:
style(eslint): fix issues raised by hound ci

vonvick commented on server/methods/core/wallet.js:
```diff
@@ -0,0 +1,76 @@
+    }
+  },
+
+  /**
```
fixed

Commit:
style(eslint): fixed more hound issues on wallet.js and paystack.js

Commit:
style: fix jsdoc issues in  wallet methods.

vonvick commented on server/methods/core/wallet.js:
```diff
@@ -0,0 +1,76 @@
+    const orderId = orderInfo._id;
+    const transaction = {amount, orderId, transactionType: "Refund"};
+    try {
+      Wallets.update({userId}, {$push: {transactions: transaction}, $inc: {balance: amount}}, {upsert: true});
```
fixed

Commit:
style: fix hound ci issues on packages publications

vonvick commented on server/publications/collections/packages.js:
```diff
@@ -37,8 +37,9 @@ Meteor.publish("Packages", function (shopCursor) {
         "layout": 1,
-        "settings.general.enabled": 1,
-        "settings.public": 1
+        "settings": 1
```
fixed

vonvick commented on server/methods/core/wallet.js:
```diff
@@ -0,0 +1,76 @@
+
+Meteor.methods({
+
+  /**
```
fixed

vonvick commented on imports/plugins/included/paystack/client/settings/paystack.js:
```diff
@@ -0,0 +1,31 @@
+      return Alerts.add("Paystack settings saved.", "success");
+    },
+    onError: function (operation, error) {
+      Alerts.removeSeen();
```
fixed

vonvick commented on imports/plugins/included/paystack/client/settings/paystack.js:
```diff
@@ -0,0 +1,31 @@
+  "paystack-update-form": {
+    onSuccess: function () {
+      Alerts.removeSeen();
+      return Alerts.add("Paystack settings saved.", "success");
```
fixed

vonvick commented on imports/plugins/included/paystack/client/settings/paystack.js:
```diff
@@ -0,0 +1,31 @@
+AutoForm.hooks({
+  "paystack-update-form": {
+    onSuccess: function () {
+      Alerts.removeSeen();
```
fixed

vonvick commented on imports/plugins/included/paystack/client/settings/paystack.js:
```diff
@@ -0,0 +1,31 @@
+  }
+});
+
+AutoForm.hooks({
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,114 @@
+
+Template.paystackPaymentForm.helpers({
+  PaystackSchema() {
+    return PaystackSchema;
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,114 @@
+    const transactionId = generateTransactionID();
+    const mailRegex = /^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,63}$/i;
+    if (!mailRegex.test(userMail)) {
+      Alerts.toast("Invalid email address", "error");
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,114 @@
+    const accountDetails = Accounts.find(Meteor.userId()).fetch();
+    const userMail = accountDetails[0].emails[0].address;
+    const amount = getOrderPrice();
+    const transactionId = generateTransactionID();
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,114 @@
+};
+
+Template.paystackPaymentForm.events({
+  "click #paywithpaystack": (event) => {
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,114 @@
+    amount: amount * 100,
+    ref: transactionId,
+    callback: function (response) {
+      handlePayment(response.reference, "payment");
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,114 @@
+// Paystack payment
+const payWithPaystack = (email, amount, transactionId) => {
+  const paystackDetails = getPaystackSettings();
+  const handler = PaystackPop.setup({
```
fixed

vonvick commented on server/methods/core/wallet.js:
```diff
@@ -0,0 +1,75 @@
+    }
+  },
+
+  /**
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+
+Template.paystackPaymentForm.helpers({
+  PaystackSchema() {
+    return PaystackSchema;
```
fixed

vonvick commented on server/methods/core/wallet.js:
```diff
@@ -0,0 +1,75 @@
+    }
+  },
+
+  /**
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+    const accountDetails = Accounts.find(Meteor.userId()).fetch();
+    const userMail = accountDetails[0].emails[0].address;
+    const amount = getOrderPrice();
+    const transactionId = generateTransactionID();
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+    amount: amount * 100,
+    ref: transactionId,
+    callback: function (response) {
+      handlePayment(response.reference, "payment");
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+
+Template.paystackPaymentForm.helpers({
+  PaystackSchema() {
+    return PaystackSchema;
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+// Paystack payment
+const payWithPaystack = (email, amount, transactionId) => {
+  const paystackDetails = getPaystackSettings();
+  const handler = PaystackPop.setup({
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+    const accountDetails = Accounts.find(Meteor.userId()).fetch();
+    const userMail = accountDetails[0].emails[0].address;
+    const amount = getOrderPrice();
+    const transactionId = generateTransactionID();
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+    amount: amount * 100,
+    ref: transactionId,
+    callback: function (response) {
+      handlePayment(response.reference, "payment");
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+          transactionId: paystackResponse.reference,
+          currency: paystackResponse.currency
+        });
+        finalizePayment(paystackMethod);
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+// Paystack payment
+const payWithPaystack = (email, amount, transactionId) => {
+  const paystackDetails = getPaystackSettings();
+  const handler = PaystackPop.setup({
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+      };
+      if (type === "payment") {
+        paystackMethod.transactions = [];
+        paystackMethod.transactions.push({
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+        createdAt: new Date()
+      };
+      if (type === "payment") {
+        paystackMethod.transactions = [];
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+          transactionId: paystackResponse.reference,
+          currency: paystackResponse.currency
+        });
+        finalizePayment(paystackMethod);
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+      };
+      if (type === "payment") {
+        paystackMethod.transactions = [];
+        paystackMethod.transactions.push({
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,115 @@
+        createdAt: new Date()
+      };
+      if (type === "payment") {
+        paystackMethod.transactions = [];
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,114 @@
+    const transactionId = generateTransactionID();
+    const mailRegex = /^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,63}$/i;
+    if (!mailRegex.test(userMail)) {
+      Alerts.toast("Invalid email address", "error");
```
fixed

vonvick commented on imports/plugins/core/layout/client/templates/layout/wallet/wallet.js:
```diff
@@ -0,0 +1,144 @@
+    headers: {
+      Authorization: `Bearer ${paystackConfig.settings.secretKey}`
+    }
+  }, function (error, response) {
```
fixed

vonvick commented on imports/plugins/included/paystack/client/checkout/paystack.js:
```diff
@@ -0,0 +1,114 @@
+          transactionId: paystackResponse.reference,
+          currency: paystackResponse.currency
+        });
+        finalizePayment(paystackMethod);
```
fixed

vonvick commented on server/methods/core/wallet.js:
```diff
@@ -0,0 +1,70 @@
+    }
+  },
+
+  /**
```
fixed

andela-aolaniran reviewed (approved):
Looks good.

Commit:
fix merge conflict while pulling from staging
Merge branch 'staging' of https://github.com/andela/anbu-rc into feature/137005167/paystack-payment-and-wallet

Commit:
refactor(wallet): fix issue with wallet to wallet transfer logic

Commit:
style: change amount from let to const
